### PR TITLE
Generation of block candidate with certain transactions included

### LINKED ISSUE #929: Candidate block generation with certain transactions included
Make an alternative to "/mining/candidate" which is also accepting transactions  to be necessarily included into blocks. The method should return block header and Merkle proofs for the transactions.

### PR DESCRIPTION
**Pull request by kushti**
Close #929 , this PR introduces an alternative to "/mining/candidateWithTxs" which is also accepting transactions to be necessarily included into blocks. The method returns block header and Merkle proofs for the transactions.

### CODE REVIEW

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala [hunk 1]
```diff
@@ -5,24 +5,55 @@ import io.circe.{Decoder, Encoder, HCursor}
...
+                             txs: Seq[ErgoTransaction],
+                             override val sizeOpt: Option[Int] = None)
+  extends BlockSection with TransactionsCarryingPersistentNodeViewModifier[ErgoTransaction] {
 
   assert(txs.nonEmpty, "Block should always contain at least 1 coinbase-like transaction")
 
   override val modifierTypeId: ModifierTypeId = BlockTransactions.modifierTypeId
 
-  override def digest: Digest32 = BlockTransactions.transactionsRoot(txs)
+  lazy val txIds: Seq[Array[Byte]] = txs.map(_.serializedId)
```
Comment (greenhat) [inline src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala]: Opportunity for `NonEmptyList`?

File: src/main/scala/org/ergoplatform/settings/Algos.scala [hunk 2]
```diff
@@ -17,11 +17,19 @@ object Algos extends ErgoAlgos with scorex.core.utils.ScorexEncoding {
...
 
+  /**
+    * A method which is building a Merkle tree over binary objects and returns digest
+    * (256-bits long root hash) of the tree
+    * @param elements - tree leafs
+    * @return 256-bits (32-bytes) long digest of the tree
+    */
   def merkleTreeRoot(elements: Seq[LeafData]): Digest32 =
-    if (elements.isEmpty) emptyMerkleTreeRoot else MerkleTree(elements)(hash).rootHash
+    if (elements.isEmpty) emptyMerkleTreeRoot else merkleTree(elements).rootHash
```
Comment (greenhat) [inline src/main/scala/org/ergoplatform/settings/Algos.scala]: How `emptyMerkleTreeRoot` is different from `EmptyRootNode.hash`, which `MerkleTree.rootHash`should return  if `elements` are empty? Also, `emptyMerkleTreeRoot` is a hash of empty byte array and `EmptyRootNode.hash` is `hf.byteArrayToDigest(Array.fill(hf.DigestSize)(0: Byte)).get` (hash of non-empty byte array).

File: src/main/scala/org/ergoplatform/local/ErgoMiner.scala [hunk 3]
```diff
@@ -486,25 +496,60 @@ object ErgoMiner extends ScorexLogging {
...
     }
 
-    loop(mempoolTxsIn, startTransactions, None, Seq.empty)
+    loop(mempoolTxsIn, mandatoryTxs, None, Seq.empty)
   }
 
-  def fixTxsConflicts(txs: Seq[(ErgoTransaction, Long)]): Seq[(ErgoTransaction, Long)] = {
-    txs.foldLeft((Seq.empty[(ErgoTransaction, Long)], Set.empty[ByteArrayWrapper])) { case ((s, keys), (tx, cost)) =>
-      val bxsBaw = tx.inputs.map(_.boxId).map(ByteArrayWrapper.apply)
+  def fixTxsConflicts(txs: Seq[CostedTransaction]): Seq[CostedTransaction] = {
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/local/ErgoMiner.scala]: replaced with a "doubleSpend" predicate

File: src/main/scala/org/ergoplatform/settings/Algos.scala (hunk 2 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/settings/Algos.scala]: Good catch! Made an issue on that: https://github.com/ergoplatform/ergo/issues/1077

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 1 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala]: @greenhat from which framework ?

**Comment by kushti** [comment]
@greenhat @aslesarenko addressed comments, updated openapi.yaml , please make another pass on it!

File: src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala (hunk 1 above)
Comment (greenhat) [reply src/main/scala/org/ergoplatform/modifiers/history/BlockTransactions.scala]: I believe we should have `cats` in dependencies. So  (`NonEmptyList`)[https://typelevel.org/cats/datatypes/nel.html] seems like a good fit.

File: src/main/scala/org/ergoplatform/mining/ProofOfUpcomingTransactions.scala [hunk 4]
```diff
@@ -0,0 +1,39 @@
+package org.ergoplatform.mining
+
+import io.circe.{Encoder, Json}
+import org.ergoplatform.modifiers.history.{BlockTransactions, HeaderSerializer, HeaderWithoutPow}
+import org.ergoplatform.nodeView.mempool.TransactionMembershipProof
+import org.ergoplatform.settings.Algos
+import io.circe.syntax._
+
+/**
+  * Proof of inclusion of certain transactions into a block with known and yet unproven header
```
Comment (aslesarenko) [inline src/main/scala/org/ergoplatform/mining/ProofOfUpcomingTransactions.scala:10]: please describe the rationale/motivation behind this design, what are the goals, etc.

File: src/main/scala/org/ergoplatform/local/ErgoMiner.scala [hunk 5]
```diff
@@ -362,6 +388,39 @@ class ErgoMiner(ergoSettings: ErgoSettings,
...
+    */
+  case class CostedTransaction(tx: ErgoTransaction, cost: Long) {
+    lazy val id: ModifierId = tx.id
+
+    override def equals(obj: Any): Boolean = obj match {
+      case c: CostedTransaction => c.id == id
+      case _ => false
+    }
+
+    override def hashCode(): Int = tx.hashCode()
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/local/ErgoMiner.scala]: Removed case class, CostedTransaction is now a type alias for the (ErgoTransaction, Long) tuple.

File: src/main/scala/org/ergoplatform/mining/ProofOfUpcomingTransactions.scala (hunk 4 above)
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/ProofOfUpcomingTransactions.scala]: A particular example is provided.

File: src/main/resources/api/openapi.yaml [hunk 6]
```diff
@@ -1025,6 +1065,66 @@ components:
...
+          items:
+            type: string
+            description: internal node with direction byte prepended
+            example: '018b7ae20a4acd23e3f1bf38671ce97103ad96d8f1c780b5e5e865e4873ae16337'
+          example: ['018b7ae20a4acd23e3f1bf38671ce97103ad96d8f1c780b5e5e865e4873ae16337']
+
+
+    ProofOfUpcomingTransactions:
+      type: object
+      description: Proof that a block corresponding to given header without PoW contains certain transactions
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/resources/api/openapi.yaml:1090]: renamed certain => some

File: src/main/scala/org/ergoplatform/mining/ExternalCandidateBlock.scala [hunk 7]
```diff
@@ -5,16 +5,34 @@ import io.circe.{Encoder, Json}
...
+/**
+  * Block candidate for external miner
+  *
+  * @param msg                            - message for external miner to work on
+  * @param b                              - target value for mining
+  * @param pk                             - public key of a miner
+  * @param proofsForMandatoryTransactions - proofs of transactions membership (optional)
+  *
+  */
+case class ExternalCandidateBlock(msg: Array[Byte],
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/mining/ExternalCandidateBlock.scala:18]: "Data" suffix not saying anything concrete, renamed into "WorkMessage" (making similar to getwork in Bitcoin)

File: src/main/scala/org/ergoplatform/local/ErgoMiner.scala [hunk 8]
```diff
@@ -362,6 +388,39 @@ class ErgoMiner(ergoSettings: ErgoSettings,
...
+  /**
+    * Holder for both candidate block and data for external miners derived from it
+    * (to avoid possibly costly recalculation)
+    *
+    * @param candidateBlock
+    * @param externalVersion
+    * @param txsToInclude
+    */
+  private case class CandidateCache(candidateBlock: CandidateBlock,
+                                    externalVersion: ExternalCandidateBlock,
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/local/ErgoMiner.scala]: What does "Data" suffix saying?

File: src/main/scala/org/ergoplatform/local/ErgoMiner.scala [hunk 9]
```diff
@@ -263,27 +279,41 @@ class ErgoMiner(ergoSettings: ErgoSettings,
...
-
+  /**
+    * Assemble correct block candidate based on
+    *
+    * @param minerPk
+    * @param history
+    * @param pool
+    * @param proposedUpdate
+    * @param state
+    * @param mandatoryTransactions
```
[the remark this answers is not in the document]
Comment (kushti) [reply src/main/scala/org/ergoplatform/local/ErgoMiner.scala]: Provided descriptions

**Comment by kushti** [comment]
@aslesarenko issues and questions are resolved / answered, please make another pass!

File: src/main/scala/org/ergoplatform/local/ErgoMiner.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (aslesarenko) [reply src/main/scala/org/ergoplatform/local/ErgoMiner.scala]: The Version part of the name is misleading. How should I read this declaration `externalVersion: WorkMessage`. Version of message? Is there non-external version?

**Pull request merged**